src/classes/RD_RecurringDonations_TEST2.cls

```diff
@@ -1084,36 +1091,48 @@ private with sharing class RD_RecurringDonations_TEST2 {
-        system.assertEquals(12, listOpp28th.size());
+        system.assertEquals(12, listOpp28th.size(), 'There should be 12 Opps with a Close Date of the 28th');
         for (Opportunity opp : listOpp28th) {
-            system.assertEquals(28, opp.CloseDate.Day());
+            system.assertEquals(28, opp.CloseDate.Day(), 'The Opp.CloseDate.Day should be the 28th');
         }
 
         // All Installments should be on the 30th except for February, which should be on the last day of the month
         list<Opportunity> listOpp30th = [select Id, CloseDate from Opportunity where npe03__Recurring_Donation__r.Name = 'test30th' order by CloseDate];
-        system.assertEquals(12, listOpp30th.size());
+        system.assertEquals(12-opps30thMonthOffset, listOpp30th.size(), 'There should be ' + (12-opps30thMonthOffset) +
+                ' Opps with a Close Date of the 30th');
         for (Opportunity opp : listOpp28th) {
             if (opp.CloseDate.Month() != 2) {
-                system.assertEquals(28, opp.CloseDate.Day());
+                system.assertEquals(28, opp.CloseDate.Day(), 'The Opp.CloseDate.Day should be the 28th');
             } else {
-                system.assertEquals(Date.DaysInMonth(opp.CloseDate.Year(), 2), opp.CloseDate.Day());
+                system.assertEquals(Date.DaysInMonth(opp.CloseDate.Year(), 2), opp.CloseDate.Day(),
+                'The installment close date for February should be on the last day of February for the year');
             }
         }
 
         // All Installments should be on the last day of the month
         list<Opportunity> listOppEnd = [select Id, CloseDate from Opportunity where npe03__Recurring_Donation__r.Name = 'testEndOfMonth' order by CloseDate];
-        system.assertEquals(12, listOppEnd.size());
+        system.assertEquals(12, listOppEnd.size(), 'There should be 12 Opps with a close date of the last day of the month');
         for (Opportunity opp : listOppEnd) {
-            system.assertEquals(Date.DaysInMonth(opp.CloseDate.Year(), opp.CloseDate.Month()), opp.CloseDate.Day(), opp.CloseDate);
+            system.assertEquals(Date.DaysInMonth(opp.CloseDate.Year(), opp.CloseDate.Month()), opp.CloseDate.Day(),
+                'The Opps Close Date should be the very last day of the month: ' + opp.CloseDate);
+        }
+
+        // For each created installment (going by the 30th), compare the month to the created installments for
+        // the 28th installments. There will always be 12 of those, but there may be less installments for the
+        // ones with a close date of the 30th if the start date was in the future.
+        for (integer i = 0; i < 12-opps30thMonthOffset; i++) {
+            Integer expectedMonth = (listOpp28th[i].CloseDate.Month() + opps30thMonthOffset > 12
+                    ? (listOpp28th[i].CloseDate.Month() + opps30thMonthOffset) - 12
```

Comment:
This section is a bit confusing, I am not sure we should have the 30th test referring to the 28th records?

Reply:
Given the root issue and complexity in this particular test, I've been hesitant to completely rewrite/split into different tests. The above loop is definitely not intuitive, but my reading of it is that the original intent appears to be to compare the generated installments for the 28th and 30th to make sure that the "Month" for each one of the matches. For example, if the RD starts this month, both should create installments from Sept through Aug (next year) - assuming 12 installments are created. Of course, the new `opps30thMonthOffset` var is used to track if the logic creates less than 12 installments and adjust the loop/logic accordingly. The goes through each installment for each RD and makes sure the month matches.

If we are willing to drop this particular verification, then it would definitely make sense to split the `testOpenEndedMonthlyEndDates` method into individual methods to test the 28, 30th and end of month. I suppose that theoretically if each individual test can validate the number of created installments, and possibly the starting and ending months of those installments, then it wouldn't actually be necessary to use a loop like this.

Reply:
I think that makes the most sense because you are really testing the end dates are the end of the month vs the month itself. It might be worth testing the end of months by just looping through your query results and ensuring that the date + 1 is equal to the next date + 1 month - 1day to assert that it's the last day of each month.

Reply:
I'll give it a shot.